server/app/auth/oidc/CiviformOidcLogoutActionBuilder.java

```diff
@@ -0,0 +1,95 @@
+import org.pac4j.core.exception.TechnicalException;
+import org.pac4j.core.exception.http.RedirectionAction;
+import org.pac4j.core.profile.UserProfile;
+import org.pac4j.core.util.CommonHelper;
+import org.pac4j.core.util.HttpActionHelper;
+import org.pac4j.oidc.config.OidcConfiguration;
+import org.pac4j.oidc.logout.OidcLogoutActionBuilder;
+
+/**
+ * Custom OidcLogoutActionBuilder for CiviFormProfileData (since it extends CommonProfile, not
+ * OidcProfile). Also allows for divergence from the [oidc
+ * spec](https://openid.net/specs/openid-connect-rpinitiated-1_0.html) if your provider requires it
+ * (e.g. Auth0).
+ *
+ * <p>Does not provide the recommended id_token_hint, since these are not stored by the civiform
+ * profile.
+ *
+ * <p>Uses the post_logout_redirect_uri parameter by default, but allows overriding to a different
+ * value using the auth.oidc_post_logout_param config variable
+ *
+ * <p>If the oidc_logout_client_id_param config variable is set, also adds the client_id to the
+ * logout request.
+ */
+public final class CiviformOidcLogoutActionBuilder extends OidcLogoutActionBuilder {
+
+  private Optional<String> postLogoutRedirectParam;
+  private ImmutableMap<String, String> extraParams;
+
+  public CiviformOidcLogoutActionBuilder(
+      Config civiformConfiguration, OidcConfiguration oidcConfiguration, String clientID) {
+    super(oidcConfiguration);
+    checkNotNull(civiformConfiguration);
+    this.postLogoutRedirectParam =
+        getConfigurationValue(civiformConfiguration, "auth.oidc_post_logout_param");
+    Optional<String> clientIdParam =
+        getConfigurationValue(civiformConfiguration, "auth.oidc_logout_client_id_param");
+
+    if (clientIdParam.isPresent()) {
+      this.extraParams = ImmutableMap.of(clientIdParam.get(), clientID);
+    } else {
+      this.extraParams = ImmutableMap.of();
+    }
+  }
+
+  /** Helper function for retriving values from the application.conf, */
+  static final Optional<String> getConfigurationValue(Config civiformConfiguration, String name) {
+    if (civiformConfiguration.hasPath(name)) {
+      return Optional.ofNullable(civiformConfiguration.getString(name));
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * Override the parent's getLogoutAction, since it checks that the profile is an instance of
+   * OidcProfile, and uses fields we don't have access to. Generally keeps the same basic logic.
+   *
+   * <p>Also use the custom CustomOidcLogoutRequest to build the URL.
+   */
+  @Override
+  public Optional<RedirectionAction> getLogoutAction(
+      WebContext context, SessionStore sessionStore, UserProfile currentProfile, String targetUrl) {
+    String logoutUrl = configuration.findLogoutUrl();
+    if (CommonHelper.isNotBlank(logoutUrl) && currentProfile instanceof CiviFormProfileData) {
+      try {
+        URI endSessionEndpoint = new URI(logoutUrl);
+        LogoutRequest logoutRequest =
+            new CustomOidcLogoutRequest(
```

Comment:
It appears that the base implementation forks based on whether target URL is blank or not (https://github.com/pac4j/pac4j/blob/master/pac4j-oidc/src/main/java/org/pac4j/oidc/logout/OidcLogoutActionBuilder.java).

Generally, I'm a little hesitant to fork code, especially in a library specifically for security. If some sort of security patch were added in the base class, we wouldn't pick it up nor would we have any way of knowing we needed to.

If we could support ID tokens in the CiviformProfile, is there any way we'd be able to use the base implementation and not require custom subclassing? As I understand, it might not be possible since we wouldn't have a way of passing additional query parameters in the callback.

If we stay with this implementation, I'd prefer we open an issue against pac4j / etc to support this specific scenario since it avoids the need to code fork.

Reply:
The basic issue here is that Auth0 (and maybe other providers) don't comply with the Oauth spec for logout.  It's not really pac4j's fault. 

However, the big problem (and bigger refactor) is that the parent uses the line `if (CommonHelper.isNotBlank(logoutUrl) && currentProfile instanceof OidcProfile)`.  We don't inherit from OidcProfile, so it's always false.

The fork you mention is not really a fork in logic.  It uses a shortcut constructor that calls the main one, but just inserts 'null' so you don't have to.

Reply:
(in order to fix the oidc profile thing, we'd need to remove all our saml code, or branch the logic somehow)

Reply:
I definitely agree that the logic now is consistent with what we want. My worry is that the underlying Pac4J library will drift over time and we'll silently fall out of sync with the implementation.

I definitely agree that the non-standards-compliant implementations complicate this. I don't think we need to block this PR over it necessarily, but I do think we might want to have a larger discussion about if there are other hooks in Pac4J that we could use to not have to fork the code. Or perhaps we can file a bug to support these scenarios, since I imagine the Pac4J maintainers likely would prefer having customization hooks to potentially forking the implementation.

Again, this concern needn't be considered blocking in any way for the PR, I'm just trying to voice my concern of drift over time.

Reply:
➕